### PR TITLE
feat: Add Ruby code - chapter "Stack & Queue"

### PR DESCRIPTION
**Pull request by khoaxuantu**
> Due to the huge load of code, the Ruby code blocks of .md files for this chapter will be in another PR.

---

If this pull request (PR) is associated with **coding or code transpilation**, please attach the relevant console outputs to the PR and complete the following checklist:

- [x] I have thoroughly reviewed the code, focusing on its formatting, comments, indentation, and file headers.
- [x] I have confirmed that the code execution outputs are consistent with those produced by the reference code (Python or Java).
- [x] The code is designed to be compatible on standard operating systems, including Windows, macOS, and Ubuntu.

---

stack_and_queue
stack_and_queue_1
stack_and_queue_2

### CODE REVIEW

File: codes/ruby/chapter_stack_and_queue/deque.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+# 初始化双向队列
+# Ruby 没有内直的双端队列，只能把 Array 当作双端队列来使用
+deque = []
+
+# 元素如队
+deque << 2
+deque << 5
+deque << 4
+# 请注意，由于是数组，Array#unshift 方法的时间复杂度为 O(n)
+deque.unshift(3)
```
Comment (khoaxuantu) [inline codes/ruby/chapter_stack_and_queue/deque.rb:18]: This method, as in this [comment](https://stackoverflow.com/a/47683752) stated, has been armotized to be `O(1)` in average, but an individual operation may be `O(n)`. So I keep the comment to be `O(n)` statement.

Comment (krahets) [reply codes/ruby/chapter_stack_and_queue/deque.rb:18]: That's fine. The worst time complexity is `O(n)`

File: codes/ruby/chapter_stack_and_queue/queue.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
+=begin
+File: queue.rb
+Created Time: 2024-04-06
+Author: Xuan Khoa Tu Nguyen (ngxktuzkai2000@gmail.com)
+=end
+
+### Driver Code ###
+
+# 初始化队列
+# Ruby 内置的队列 （Thread::Queue) 没有 peek 和遍历方法，可以把 Array 当作队列来使用
```
Comment (krahets) [inline codes/ruby/chapter_stack_and_queue/queue.rb:10]: Redundant space between `队列 （`

File: codes/ruby/chapter_stack_and_queue/queue.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+queue.push(5)
+queue.push(4)
+puts "队列 queue = #{queue}"
+
+# 访问队列元素
+peek = queue.first
+puts "队首元素 peek = #{peek}"
+
+# 元素出队
+# 清注意，由于是数组，Array#shift 方法时间复杂度为 O(n)
```
Comment (krahets) [inline codes/ruby/chapter_stack_and_queue/queue.rb:26]: Is `Array#shift` a standard way to denote the class method?

Comment (khoaxuantu) [reply codes/ruby/chapter_stack_and_queue/queue.rb:26]: It is a standard way to denote the instance method in Ruby.
In Ruby API documentation, the `Class#method` format represents the instance method and the `Class.method` represents the class method (static method)

For example:
https://www.rubydoc.info/stdlib/core/Array:shift

**Review by krahets — APPROVED** [review]
Thanks for the PR!

**Pull request merged**